From d7c15031623b4a1be3419177f9fb6197eec8d9a7 Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Sat, 3 Dec 2022 13:54:57 +0100
Subject: [PATCH] Add compatibility defines for libdvdnav 5.0

---
 ddvdlib.h | 12 ++++++------
 main.c    | 31 ++++++++++++++++++++++++-------
 2 files changed, 30 insertions(+), 13 deletions(-)

diff --git a/ddvdlib.h b/ddvdlib.h
index f96b9a8..9aa1275 100644
--- a/ddvdlib.h
+++ b/ddvdlib.h
@@ -157,25 +157,25 @@ void ddvd_get_last_string(struct ddvd*pconfig, void *text);
 // int id -> logical track number
 // uint16_t lang -> audio language in 2 letter iso code
 // int type -> audio type, see audio type enum (ac3,mpeg,...)
-void ddvd_get_last_audio(struct ddvd*pconfig, void *id, void *lang, void *type);
+void ddvd_get_last_audio(struct ddvd*pconfig, int *id, int16_t *lang, int *type);
 
 // get audio track details for given audio track id
-void ddvd_get_audio_byid(struct ddvd *pconfig, int audio_id, void *lang, void *type);
+void ddvd_get_audio_byid(struct ddvd *pconfig, int audio_id, int16_t *lang, int *type);
 
 // get the number of available audio tracks
-void ddvd_get_audio_count(struct ddvd *pconfig, void *count);
+void ddvd_get_audio_count(struct ddvd *pconfig, int *count);
 
 // get the active subtitle track
 // int id -> logical track number
 // uint16_t lang -> subtitle language in 2 letter iso code
 // id=-1 means no subtitle track active
-void ddvd_get_last_spu(struct ddvd*pconfig, void *id, void *lang);
+void ddvd_get_last_spu(struct ddvd*pconfig, int *id, int16_t *lang);
 
 // get track details for given subtitle track id
-void ddvd_get_spu_byid(struct ddvd *pconfig, int spu_id, void *lang);
+void ddvd_get_spu_byid(struct ddvd *pconfig, int spu_id, int16_t *lang);
 
 // get the number of available subtitle tracks
-void ddvd_get_spu_count(struct ddvd *pconfig, void *count);
+void ddvd_get_spu_count(struct ddvd *pconfig, int *count);
 
 // get dvd title string
 void ddvd_get_title_string(struct ddvd*pconfig, char *title_string);
diff --git a/main.c b/main.c
index 55c4fa3..0623b71 100644
--- a/main.c
+++ b/main.c
@@ -32,6 +32,13 @@
 #include "string.h"
 #include "errno.h"
 
+#if !defined(DVDNAV_FORMAT_AC3)
+#define DVDNAV_FORMAT_AC3      DVD_AUDIO_FORMAT_AC3
+#define DVDNAV_FORMAT_MPEGAUDIO        DVD_AUDIO_FORMAT_MPEG2_EXT
+#define DVDNAV_FORMAT_LPCM     DVD_AUDIO_FORMAT_LPCM
+#define DVDNAV_FORMAT_DTS      DVD_AUDIO_FORMAT_DTS
+#endif
+
 #define Debug(level, str, ...) (DebugLevel > level ? printf("LIBDVD: %07.3f: " str, (float) ddvd_get_time() / 1000.0, ##__VA_ARGS__) : 0)
 #define Perror(msg)            Debug(-1, "%s: %s", msg, strerror(errno))
 
@@ -413,7 +420,7 @@ void ddvd_get_last_string(struct ddvd *pconfig, void *text)
 }
 
 // get the number of available audio tracks
-void ddvd_get_audio_count(struct ddvd *pconfig, void *count)
+void ddvd_get_audio_count(struct ddvd *pconfig, int *count)
 {
 	int c = 0;
 	int i;
@@ -425,7 +432,7 @@ void ddvd_get_audio_count(struct ddvd *pconfig, void *count)
 }
 
 // get the active audio track
-void ddvd_get_last_audio(struct ddvd *pconfig, void *id, void *lang, void *type)
+void ddvd_get_last_audio(struct ddvd *pconfig, int *id, int16_t *lang, int *type)
 {
 	memcpy(id, &pconfig->last_audio_id, sizeof(pconfig->last_audio_id));
 	memcpy(lang, &pconfig->last_audio_lang, sizeof(pconfig->last_audio_lang));
@@ -434,7 +441,7 @@ void ddvd_get_last_audio(struct ddvd *pconfig, void *id, void *lang, void *type)
 }
 
 // get audio track details for given audio track id
-void ddvd_get_audio_byid(struct ddvd *pconfig, int audio_id, void *lang, void *type)
+void ddvd_get_audio_byid(struct ddvd *pconfig, int audio_id, int16_t *lang, int *type)
 {
 	int audio_id_logical;
 	uint16_t audio_lang = 0xFFFF;
@@ -448,7 +455,7 @@ void ddvd_get_audio_byid(struct ddvd *pconfig, int audio_id, void *lang, void *t
 }
 
 // get the active SPU track
-void ddvd_get_last_spu(struct ddvd *pconfig, void *id, void *lang)
+void ddvd_get_last_spu(struct ddvd *pconfig, int *id, int16_t *lang)
 {
 	memcpy(id, &pconfig->last_spu_id, sizeof(pconfig->last_spu_id));
 	memcpy(lang, &pconfig->last_spu_lang, sizeof(pconfig->last_spu_lang));
@@ -456,7 +463,7 @@ void ddvd_get_last_spu(struct ddvd *pconfig, void *id, void *lang)
 }
 
 // get the number of available subtitle tracks
-void ddvd_get_spu_count(struct ddvd *pconfig, void *count)
+void ddvd_get_spu_count(struct ddvd *pconfig, int *count)
 {
 	int c = 0;
 	int i;
@@ -469,7 +476,7 @@ void ddvd_get_spu_count(struct ddvd *pconfig, void *count)
 }
 
 // get language details for given subtitle track id
-void ddvd_get_spu_byid(struct ddvd *pconfig, int spu_id, void *lang)
+void ddvd_get_spu_byid(struct ddvd *pconfig, int spu_id, int16_t *lang)
 {
 	uint16_t spu_lang = 0xFFFF;
 	if (spu_id < MAX_SPU && pconfig->spu_map[spu_id].logical_id > -1)
@@ -590,6 +597,11 @@ static int readApiSize(int fd, int *xres, int *yres, int *aspect)
 	return -1;
 }
 
+/* The definition of VIDEO_GET_FRAME_RATE was removed in 4.18 */
+#ifndef VIDEO_GET_FRAME_RATE
+#define VIDEO_GET_FRAME_RATE       _IOR('o', 56, unsigned int)
+#endif
+
 static int readApiFrameRate(int fd, int *framerate)
 {
 	unsigned int frate;
@@ -2290,6 +2302,8 @@ send_message:
 					case DDVD_SKIP_BWD:
 					case DDVD_SET_TITLE:
 					case DDVD_SET_CHAPTER:
+					case DDVD_SET_AUDIO:
+					case DDVD_SET_SUBTITLE:
 						// we must empty the pipe here... and fall through
 						ddvd_readpipe(key_pipe, &keydone, sizeof(int), 1);
 					default:
@@ -2712,9 +2726,12 @@ err_dvdnav:
 		Debug(1, "Error on dvdnav_close: %s\n", dvdnav_err_to_string(dvdnav));
 
 err_dvdnav_open:
-	ddvd_device_clear();
+	if (ioctl(ddvd_fdvideo, VIDEO_STOP) < 0)
+		Perror("VIDEO_STOP");
 	if (ioctl(ddvd_fdvideo, VIDEO_SELECT_SOURCE, VIDEO_SOURCE_DEMUX) < 0)
 		Perror("VIDEO_SELECT_SOURCE");
+	if (ioctl(ddvd_fdvideo, AUDIO_STOP) < 0)
+		Perror("VIDEO_STOP");
 	if (ioctl(ddvd_fdaudio, AUDIO_SELECT_SOURCE, AUDIO_SOURCE_DEMUX) < 0)
 		Perror("AUDIO_SELECT_SOURCE");
 	close(ddvd_ac3_fd);
-- 
2.36.0.windows.1

